mlir/lib/Conversion/MQTRefToQIR/MQTRefToQIR.cpp

```diff
@@ -0,0 +1,881 @@
+
+  /**
+   * @brief Makes sure that the different blocks for the base profile of QIR
+   * exist
+   *
+   * The first block should only contain constant operations for the initialize
+   * operation. The second block contains the reversible quantum operations. The
+   * third block contains the irreversible quantum operations like the measure
+   * operation or the dealloc operation. The final block contains the
+   * return operation and the record output operation. The blocks are connected
+   * with an unconditional jump operation to the next block.
+   *
+   * @param main The main function of the module.
+   */
+  static void ensureBlocks(LLVM::LLVMFuncOp& main) {
+    // return if there are more blocks already
+    if (main.getBlocks().size() > 1) {
+      return;
+    }
+    // get the existing block
+    auto* entryBlock = &main.front();
+    OpBuilder builder(main.getBody());
+
+    // create the main blocks and the endblock
+    Block* mainBlock = builder.createBlock(&main.getBody());
+    Block* irreversibleMainBlock = builder.createBlock(&main.getBody());
+    Block* endBlock = builder.createBlock(&main.getBody());
+
+    // move the returnOp from the entryBlock to the endBlock
+    auto& entryOperations = entryBlock->getOperations();
+    auto& endOperations = endBlock->getOperations();
+    endOperations.splice(endOperations.end(), entryOperations,
+                         std::prev(entryOperations.end()));
+
+    // get all quantum operations until the measure operation
+    SmallVector<Operation*> opsToMove;
+    for (auto& op : entryOperations) {
+      if (dyn_cast<ref::MeasureOp>(op)) {
+        break;
+      }
+      opsToMove.emplace_back(&op);
+    }
+    // move them to the 2nd block
+    for (Operation* op : opsToMove) {
+      op->moveBefore(mainBlock, mainBlock->end());
+    }
+
+    // move the remaining operations to the third block
+    if (!entryBlock->empty()) {
+      irreversibleMainBlock->getOperations().splice(
+          irreversibleMainBlock->begin(), entryOperations,
+          entryOperations.begin(), entryOperations.end());
+    }
+
+    // add jump from entryBlock to mainBlock
+    builder.setInsertionPointToEnd(entryBlock);
+    builder.create<LLVM::BrOp>(main->getLoc(), mainBlock);
+
+    // add jump from main to irreversibleMain
+    builder.setInsertionPointToEnd(mainBlock);
+    builder.create<LLVM::BrOp>(main->getLoc(), irreversibleMainBlock);
+
+    // add jump from irreversibleMain to endBlock
+    builder.setInsertionPointToEnd(irreversibleMainBlock);
+    builder.create<LLVM::BrOp>(main->getLoc(), endBlock);
+  }
+
+  /**
+   * @brief Adds the initialize operation to the first block of the main
+   * function.
+   *
+   * @param main The main function of the module.
+   * @param ctx The context of the module.
+   */
+  static void addInitialize(LLVM::LLVMFuncOp& main, MLIRContext* ctx,
+                            LoweringState* state) {
+    auto module = main->getParentOfType<ModuleOp>();
```

Comment:
I am pretty sure it is not wise to use `module` as a name here as this is actually a reserved keyword starting with C++20